File: python3/range_sum_of_bst.py
from .model import TreeNode

"""
Space   : O(n)
Time    : O(n)
"""

class Solution:
    def rangeSumBST(self, root: TreeNode, low: int, high: int) -> int:
        if not root:
            return 0

        ans = 0
        stack = [root]

        while stack:
            node = stack.pop()
            if low <= node.val <= high:
                ans += node.val
            if node.left:
                stack.append(node.left)
            if node.right:
                stack.append(node.right)

        return ans

File: golang/range_sum_of_bst.go
package main

/*
Space	: O(n)
Time	: O(n)
*/

import "golang/model"

func rangeSumBST(root *model.TreeNode, low int, high int) int {
	var ans int
	stack := []*model.TreeNode{root}

	for len(stack) > 0 {
		temp := []*model.TreeNode{}
		for _, node := range stack {
			if node.Val >= low && node.Val <= high {
				ans += node.Val
			}
			if node.Left != nil {
				temp = append(temp, node.Left)
			}
			if node.Right != nil {
				temp = append(temp, node.Right)
			}
		}
		stack = temp
	}

	return ans
}
